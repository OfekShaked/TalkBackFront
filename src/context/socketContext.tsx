import {io} from "socket.io-client";
import React from 'react';
import config from '../configs/config';
const SOCKET_URL = config.apiUrl;

export const socket = io(SOCKET_URL, { transports: ['websocket', 'polling', 'flashsocket'] });
export const SocketContext = React.createContext<any>(null);
