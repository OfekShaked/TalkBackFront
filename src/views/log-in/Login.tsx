import React,{useState,useContext} from 'react';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import TextField from '@material-ui/core/TextField';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';
import useStyles from './LoginStyle';
import { useTheme } from '@material-ui/core/styles';
import useInput from '../../hooks/useInput';
import {login} from '../../services/auth.service'
import { useHistory } from "react-router-dom";
import {SocketContext} from '../../context/socketContext';
import {handleError} from '../../services/errorHandling.service';

const Login = (props:any) => {
    const socket = useContext(SocketContext);
    const theme = useTheme();
    const classes = useStyles(theme);
    const history = useHistory();

    const usernameRecieved = useInput('');
    const password = useInput('');
    const [credentialError,setCredentialError] = useState("")

    const onSubmit = async (e:any) => {
        try{
        e.preventDefault();
        const userData:any = { "username":usernameRecieved.value, "password":password.value};        
        if(await login(userData,setCredentialError)){
            await socket.emit("user_online",userData.username);
            props.setIsLoggedIn(true);
            history.push("/contact")
        }
    }catch(error){
        handleError(error)
    }
    }
    return (
        <Container component="main" maxWidth="xs">
            <CssBaseline />
            <div className={classes.paper}>
                <Avatar className={classes.avatar}>
                    <LockOutlinedIcon />
                </Avatar>
                <Typography component="h1" variant="h5">
                    Sign In
                </Typography>
                <form className={classes.form} method="post" onSubmit={onSubmit} noValidate>
                    <Grid container spacing={2}>
                        <Grid item xs={12}>
                            <TextField
                                variant="outlined"
                                required
                                fullWidth
                                id="username"
                                label="Username"
                                name="username"
                                autoComplete="string"
                                {...usernameRecieved}
                                error={credentialError!==""}
                                helperText={credentialError}
                            />
                        </Grid>
                        <Grid item xs={12}>
                            <TextField
                                variant="outlined"
                                required
                                fullWidth
                                name="password"
                                label="Password"
                                type="password"
                                id="password"
                                autoComplete="current-password"
                                {...password}
                                error={credentialError!==""}
                                helperText={credentialError}
                            />
                        </Grid>
                    </Grid>
                    <Button
                        type="submit"
                        fullWidth
                        variant="contained"
                        color="primary"
                        className={classes.submit}
                    >
                        Log In
                    </Button>
                </form>
            </div>
        </Container>
    );
}

export default Login;