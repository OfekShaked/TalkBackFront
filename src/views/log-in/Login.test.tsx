import React from 'react';
import renderer from 'react-test-renderer';
import Login from './Login';
import { SocketContext } from '../../context/socketContext';
import SocketMock from 'socket.io-mock';
import { login } from '../../services/auth.service'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';

configure({adapter: new Adapter()});

jest.mock('../../services/auth.service', () => {
    return {
        login: jest.fn((userData:any) => {
            return true;
        })
    }
});


let socket = new SocketMock()
const isLoggedIn = jest.fn();

describe("login component", () => {
    it('renders login page', () => {
        const component = renderer.create(
            <Login setIsLoggedIn={null}></Login>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('log in page inputs to start empty', () => {
        const component = renderer.create(
            <SocketContext.Provider value={socket.socketClient}>
                <Login setIsLoggedIn={isLoggedIn}></Login>
            </SocketContext.Provider>,
        );
        const testInstance = component.root;
        const usernameInput = testInstance.findAllByType('input')[0];
        const passwordInput = testInstance.findAllByType('input')[1];
        expect(usernameInput.props.value).toEqual('');
        expect(passwordInput.props.value).toEqual('');
    });

    it('log in succesfull', () => {

        socket.on("user_online",async(username:any)=>{
            expect(username).toBe("");
        })

        const wrapper = mount(
            <SocketContext.Provider value={socket.socketClient}>
                <Login setIsLoggedIn={isLoggedIn}></Login>
            </SocketContext.Provider>
        );
        const loginWrapper = wrapper.find('Login');
        expect(loginWrapper.length).toBe(1);
        const form = loginWrapper.find('form');
        form.simulate('submit');
        expect(login).toBeCalled();
    });

})

