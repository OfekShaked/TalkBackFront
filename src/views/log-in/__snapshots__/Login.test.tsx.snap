// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component renders login page 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    className="makeStyles-paper-1"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circular makeStyles-avatar-2 MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <h1
      className="MuiTypography-root MuiTypography-h5"
    >
      Sign In
    </h1>
    <form
      className="makeStyles-form-3"
      method="post"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="username"
              id="username-label"
            >
              Username
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="string"
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="username"
                name="username"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-7"
                >
                  <span>
                    Username
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="password"
              id="password-label"
            >
              Password
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="current-password"
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="password"
                name="password"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="password"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-7"
                >
                  <span>
                    Password
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Log In
        </span>
      </button>
    </form>
  </div>
</main>
`;
